posts/formalizing-isocrystals-in-lean.md

```diff
@@ -0,0 +1,211 @@
+
+# The formalization process
+
+Rob is a logician and Heather is a geometer, and this theorem was just a quick proof of concept for a larger project, [our paper with Frédéric Dupuis on semilinear maps](https://arxiv.org/abs/2202.05360). Rob and Johan Commelin (who provided a lot of helpful advice on this formalization) worked together on the initial mathlib development of Witt vectors. But this was Rob's first foray into the subject on his own, and Heather's first foray at all. 
+
+To quote from Rob and Johan's paper, "Witt vectors have a reputation among mathematicians of being forbidding and impenetrable." And indeed this was our experience in multiple ways. 
+Not only was it difficult to learn to work with Witt vectors; 
+what's more, this barrier to entry means that there are few elementary expositions of mathematical subjects (such as isocrystals) that depend on them.
+We couldn't find the one-dimensional isocrystal classification, worked through as above, written down anywhere. 
+
+It's a cliché to say that the name "proof assistant" is a misnomer. Usually, formalizing a mathematical result is strictly harder than proving it in the traditional manner -- the computer just gets in the way. But we had the opposite experience proving this theorem. Working out the details of this argument on paper was difficult, and we were rarely sure that we were understanding a given statement correctly, or even that it was well formed. Seeing the confirmation from Lean that a theorem statement typechecked -- and the back and forth with the system to write a type-correct statement -- greatly helped us understand the mathematics. 
+
+And the proof-checking helped too. We mentioned above a wrinkle when proving the "multiplication lemma": some coefficients of the polynomial we were interested in vanish in characteristic $p$, so we had to work first with integer polynomials and later, after canceling a power of $p$, interpret them in the field $k$. This subtlety escaped us the first time around! 
+
+![Oops...](/images/isocrystal-oops.png)
+
+Back to our original motivation: the semilinear maps machinery worked great. As we had hoped, it sat invisibly in the background. Lean's extensible syntax was easy to use to set up notation <code style="white-space:nowrap;">M₁ ≃ᶠˡ[p, k] M₂</code> for the type of Frobenius-semilinear isomorphisms from `M₁` to `M₂`.
+
+```lean 
+notation M₁ ` ≃ᶠˡ[` p `,` k `] ` M₂ :=
+  linear_equiv (witt_vector.fraction_ring.frobenius_ring_hom p k) M₁ M₂
+``` 
+
+From there, an isocrystal could be defined exactly as on paper: a $K$-module equipped with a Frobenius-semilinear automorphism.
+
+```lean
+class isocrystal (V : Type*) [add_comm_group V] 
+  extends module K(p, k) V :=
+( frob : V ≃ᶠˡ[p, k] V )
+```
+
+Check out the final statement of the classification theorem in mathlib 
+[here](https://github.com/leanprover-community/mathlib/blob/lean-3.44.1/src/ring_theory/witt_vector/isocrystal.lean#L163).
```

Comment:
I think it would be wise to use a permalink here that includes a commit hash, just in case we ever patch old lean branches